File: problems/LeetCode/TopInterview150/minSubArrayLen.ts
function minSubArrayLen(target: number, nums: number[]): number {
  let windowStart = 0;
  let ans = Infinity;

  let count = 0;

  for (let windowEnd = 0; windowEnd < nums.length; windowEnd++) {
    count += nums[windowEnd];

    while (count >= target) {
      ans = Math.min(ans, windowEnd - windowStart + 1);
      count -= nums[windowStart];
      windowStart++;
    }
  }

  return ans === Infinity ? 0 : ans;
}

// Time complexity: O(n)
// Space complexity: O(1)

function minSubArrayLen2(target: number, nums: number[]): number {
  const n = nums.length;
  let ans = Infinity;

  const prefix = new Array(n + 1).fill(0);

  for (let i = 1; i < prefix.length; i++) {
    prefix[i] = nums[i - 1] + prefix[i - 1];
  }

  function lowerBound(target: number) {
    let low = 0;
    let high = prefix.length - 1;
    let ans = prefix.length;

    while (low <= high) {
      const mid = low + Math.floor((high - low) / 2);

      if (prefix[mid] >= target) {
        high = mid - 1;
        ans = mid;
      } else {
        low = mid + 1;
      }
    }

    return ans;
  }

  for (let i = 0; i < nums.length; i++) {
    const targetNum = prefix[i] + target;
    const l = lowerBound(targetNum);

    if (l < prefix.length) {
      ans = Math.min(ans, l - i);
    }
  }

  return ans === Infinity ? 0 : ans;
}

// Time complexity: O(nlog(n))
// Space complexity: O(n)

File: problems/LeetCode/TopInterview150/minSubArrayLen.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:

        min_length = len(nums) + 1

        left = 0
        current_sum = 0

        for right, num in enumerate(nums):
            current_sum += num

            while current_sum >= target:
                min_length = min(min_length, right - left + 1)
                current_sum -= nums[left]
                left += 1

        return 0 if min_length == len(nums) + 1 else min_length


# Time: O(n)
# Space: O(1)


class Solution2:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:

        min_length = len(nums) + 1
        prefix_sum = [0]

        for num in nums:
            prefix_sum.append(prefix_sum[-1] + num)

        for left in range(len(prefix_sum)):
            required_sum = prefix_sum[left] + target
            right = bisect_left(prefix_sum, required_sum)

            if right < len(prefix_sum):
                min_length = min(min_length, right - left)

        return 0 if min_length == len(nums) + 1 else min_length


# Time: O(n*log(n))
# Space: O(n)